tamr_client/transformations.py

```diff
@@ -0,0 +1,150 @@
+    """Convert input transformations to JSON data (serialize)
+
+    Args:
+        tx: Input transformation to convert
+    """
+    # datasetId omitted, only one of "datasetId" or "relativeDatasetId" is required
+    dataset_json = [
+        {"name": d.name, "relativeDatasetId": d.url.path} for d in tx.datasets
+    ]
+
+    return {"datasets": dataset_json, "transformation": tx.transformation}
+
+
+def _to_json(tx: Transformations) -> JsonDict:
+    """Convert transformations to JSON data (serialize)
+
+    Args:
+        tx: Transformations to convert
+    """
+    return {
+        "parameterized": [_input_transformation_to_json(t) for t in tx.input_scope],
+        "unified": tx.unified_scope,
+    }
+
+
+def get_all(session: Session, project: Project) -> Transformations:
+    """Get the transformations of a Project
+
+    Args:
+        project: Project containing transformations
+
+    Raises:
+        requests.HTTPError: If any HTTP error is encountered.
+
+    Example:
+        >>> import tamr_client as tc
+        >>> session = tc.session.from_auth('username', 'password')
+        >>> instance = tc.instance.Instance(host="localhost", port=9100)
+        >>> project1 = tc.project.from_resource_id(session, instance, id='1')
+        >>> print(tc.transformations.get_all(session, project1))
+    """
+    r = session.get(
+        f"{project.url}/transformations",
+        headers={"Content-Type": "application/json", "Accept": "application/json"},
+    )
+    response.successful(r)
+    return _from_json(session, project.url.instance, r.json())
+
+
+def replace_all(
+    session: Session, project: Project, tx: Transformations
+) -> requests.Response:
+    """Replaces the transformations of a Project
+
+    Args:
+        project: Project to place transformations within
+        tx: Transformations to put into project
+
+    Raises:
+        requests.HTTPError: If any HTTP error is encountered.
+
+    Example:
+        >>> import tamr_client as tc
+        >>> session = tc.session.from_auth('username', 'password')
+        >>> instance = tc.instance.Instance(host="localhost", port=9100)
+        >>> project1 = tc.project.from_resource_id(session, instance, id='1')
+        >>> dataset3 = tc.dataset.from_resource_id(session, instance, id='3')
+        >>> new_input_tx = tc.InputTransformation("SELECT *, upper(name) as name;", [dataset3])
+        >>> all_tx = tc.Transformations(
+        ... input_scope=[new_input_tx],
+        ... unified_scope=["SELECT *, 1 as one;"]
+        ... )
+        >>> tc.transformations.replace_all(session, project1, all_tx)
+    """
+    body = _to_json(tx)
+    r = session.put(
+        f"{project.url}/transformations",
+        json=body,
+        headers={"Content-Type": "application/json", "Accept": "application/json"},
+    )
+    if r.status_code == 400:
+        try:
+            r_json = r.json()
+            if r_json["class"] == "java.lang.IllegalArgumentException":
```

Comment:
How sure are we that this error class will always correspond to an invalid input dataset? Same goes for the `javax.ws.rs.BadRequestException`. Could that be caused by something other than linting?

In general, we've interpreted HTTP error codes as custom errors, but that's mostly because those errors are consistent at the protocol-level, whereas this depends on the implementation of our Tamr server (which changes much more rapidly than the HTTP spec 😛 ).

Reply:
@pcattori  Yeah I was debating this. I'm not that sure that these will be stable, is this something we can ask a DEV about? Two very different and common errors come back as code 400. Without it, we need to leave these as generic expectations.

Reply:
This is a place where precision matters, so could we check the stacktrace for more info to determine if this is indeed the common exception we are expecting?

Reply:
LintingErrors should be very common. Invalid dataset is probably a rare one. The stacktrace on the LintingError still feels a bit generic, not sure what piece to focus on. What do you think of doing a regex match on the message itself?